File: crimson-zig/src/typo/spawns.zig
const std = @import("std");
const native_math = @import("../runtime/native_math.zig");
const spawn_mod = @import("../runtime/spawn.zig");

pub const TypoSpawnCall = struct {
    pos_x: f32,
    pos_y: f32,
    type_id: spawn_mod.CreatureTypeId,
    tint_r: f32,
    tint_g: f32,
    tint_b: f32,
};

pub const TypoSpawnBatch = struct {
    cooldown_ms: i32,
    count: usize = 0,
    calls: [8]TypoSpawnCall = [_]TypoSpawnCall{undefined} ** 8,

    pub fn slice(self: *const TypoSpawnBatch) []const TypoSpawnCall {
        return self.calls[0..self.count];
    }
};

fn clamp01(value: f32) f32 {
    return std.math.clamp(value, @as(f32, 0.0), @as(f32, 1.0));
}

pub fn tickTypoSpawns(
    elapsed_ms: i32,
    spawn_cooldown_ms: i32,
    frame_dt_ms: i32,
    player_count: i32,
    world_width: f32,
    world_height: f32,
) TypoSpawnBatch {
    var batch: TypoSpawnBatch = .{
        .cooldown_ms = spawn_cooldown_ms,
    };
    const safe_player_count = @max(1, player_count);
    batch.cooldown_ms -= frame_dt_ms * safe_player_count;

    while (batch.cooldown_ms < 0 and batch.count + 2 <= batch.calls.len) {
        batch.cooldown_ms += 3500 - @divTrunc(elapsed_ms, 800);
        batch.cooldown_ms = @max(100, batch.cooldown_ms);

        // typo_gameplay_update_and_render (0x00445af4..0x00445c15): float
        // literals at PC24; fsin/fcos stay wide until the next op rounds.
        const tint_t = @as(f32, @floatFromInt(elapsed_ms + 1));
        const tint_r = clamp01(native_math.pc24Add(native_math.pc24Mul(tint_t, @as(f32, 0.00000833333343)), @as(f32, 0.3)));
        const tint_g = clamp01(native_math.pc24Add(native_math.pc24Mul(tint_t, @as(f32, 10000.0)), @as(f32, 0.3)));
        const tint_b = clamp01(native_math.pc24Add(
            @sin(@as(f64, native_math.pc24Mul(tint_t, @as(f32, 0.000100000005)))),
            @as(f32, 0.3),
        ));

        const t = native_math.pc24Mul(@as(f32, @floatFromInt(elapsed_ms)), @as(f32, 0.001));
        const y = native_math.pc24Add(
            native_math.pc24Mul(@cos(@as(f64, t)), @as(f32, 256.0)),
            native_math.pc24Mul(world_height, @as(f32, 0.5)),
        );

        batch.calls[batch.count] = .{
            .pos_x = native_math.pc24Add(world_width, @as(f32, 64.0)),
            .pos_y = y,
            .type_id = .spider_sp2,
            .tint_r = tint_r,
            .tint_g = tint_g,
            .tint_b = tint_b,
        };
        batch.count += 1;
        batch.calls[batch.count] = .{
            .pos_x = -64.0,
            .pos_y = y,
            .type_id = .alien,
            .tint_r = tint_r,
            .tint_g = tint_g,
            .tint_b = tint_b,
        };
        batch.count += 1;
    }

    return batch;
}

test "typo spawns produce left/right pair" {
    const batch = tickTypoSpawns(1000, 0, 16, 1, 1024.0, 1024.0);
    try std.testing.expectEqual(@as(usize, 2), batch.count);
    try std.testing.expect(batch.calls[0].pos_x > 1024.0);
    try std.testing.expect(batch.calls[1].pos_x < 0.0);
}

File: src/crimson/typo/spawns.py
from __future__ import annotations

import math

import msgspec

from grim.color import RGBA
from grim.geom import Vec2
from grim.math import clamp01

from ..creatures.spawn import CreatureTypeId
from ..math_parity import f32, x87_pc24_add, x87_pc24_cos_mul, x87_pc24_mul


class TypoSpawnCall(msgspec.Struct, frozen=True):
    pos: Vec2
    type_id: CreatureTypeId
    tint_rgba: RGBA


def tick_typo_spawns(
    *,
    elapsed_ms: int,
    spawn_cooldown_ms: int,
    frame_dt_ms: int,
    player_count: int,
    world_width: float,
    world_height: float,
) -> tuple[int, list[TypoSpawnCall]]:
    elapsed_ms = int(elapsed_ms)
    cooldown = int(spawn_cooldown_ms)
    dt_ms = int(frame_dt_ms)
    player_count = max(1, int(player_count))

    cooldown -= dt_ms * player_count

    spawns: list[TypoSpawnCall] = []
    while cooldown < 0:
        cooldown += 3500 - elapsed_ms // 800
        cooldown = max(100, cooldown)

        # `typo_gameplay_update_and_render` (0x00445af4..0x00445c15): float
        # literals at PC24; `fsin`/`fcos` stay wide until the next op rounds.
        tint_t = float(elapsed_ms + 1)
        tint_r = clamp01(x87_pc24_add(x87_pc24_mul(tint_t, f32(0.00000833333343)), f32(0.3)))
        tint_g = clamp01(x87_pc24_add(x87_pc24_mul(tint_t, 10000.0), f32(0.3)))
        tint_b = clamp01(x87_pc24_add(math.sin(x87_pc24_mul(tint_t, f32(0.000100000005))), f32(0.3)))
        tint = RGBA(tint_r, tint_g, tint_b, 1.0)

        t = x87_pc24_mul(float(elapsed_ms), f32(0.001))
        y = x87_pc24_add(x87_pc24_cos_mul(t, 256.0), x87_pc24_mul(float(world_height), 0.5))

        spawns.append(
            TypoSpawnCall(
                pos=Vec2(x87_pc24_add(float(world_width), 64.0), y),
                type_id=CreatureTypeId.SPIDER_SP2,
                tint_rgba=tint,
            ),
        )
        spawns.append(
            TypoSpawnCall(
                pos=Vec2(-64.0, y),
                type_id=CreatureTypeId.ALIEN,
                tint_rgba=tint,
            ),
        )

    return cooldown, spawns
